cmd/kube-proxy/app/server.go

```diff
@@ -628,7 +638,49 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
 		return nil, err
 	}
 
-	rawNodeIPs := getNodeIPs(ctx, s.Client, s.Hostname)
+	noProxyName, err := labels.NewRequirement(apis.LabelServiceProxyName, selection.DoesNotExist, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	noHeadlessEndpoints, err := labels.NewRequirement(v1.IsHeadlessService, selection.DoesNotExist, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*noProxyName, *noHeadlessEndpoints)
+
+	// Make informers that filter out objects that want a non-default service proxy.
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(s.Client, s.Config.ConfigSyncPeriod.Duration,
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = labelSelector.String()
+		}))
+
+	s.serviceInformer = informerFactory.Core().V1().Services()
+	s.endpointSliceInformer = informerFactory.Discovery().V1().EndpointSlices()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+		s.serviceCIDRInformer = informerFactory.Networking().V1alpha1().ServiceCIDRs()
+	}
+	informerFactory.Start(wait.NeverStop)
+
+	// Make an informer that selects for our nodename.
+	currentNodeInformerFactory := informers.NewSharedInformerFactoryWithOptions(s.Client, s.Config.ConfigSyncPeriod.Duration,
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", s.NodeRef.Name).String()
+		}))
+	currentNodeInformerFactory.Start(wait.NeverStop)
+	s.nodeInformer = currentNodeInformerFactory.Core().V1().Nodes()
+
+	if !cache.WaitForNamedCacheSync("kube-proxy informers", ctx.Done(),
+		s.nodeInformer.Informer().HasSynced,
+		s.serviceInformer.Informer().HasSynced,
+		s.endpointSliceInformer.Informer().HasSynced) {
+		return nil, fmt.Errorf("kube-proxy informers could not start")
+	}
+
+	rawNodeIPs := getNodeIPs(ctx, s.nodeInformer.Lister(), s.Hostname)
```

Comment:
Does informer guarantee the sync-ed data is up to update with etcd rather than just API server cache?

Reply:
informer goes to watch cache, there is an event handler that runs later that detects if the node is stale and restart the apiserver as the Get() can also present problems wtih spot vms per example